File: chatbot/backend-python/src/llm/custom.py
"""
Custom OpenAI-compatible provider.

Supports any API that follows the OpenAI chat completions format, including:
- Ollama (http://localhost:11434/v1)
- vLLM
- LocalAI
- Azure OpenAI
- Together AI
"""

from typing import AsyncGenerator, List

from openai import AsyncOpenAI

from .interface import LLMProvider, ChatMessage
from ..config import LLMConfig


class CustomProvider(LLMProvider):
    def __init__(self, config: LLMConfig):
        self._config = config
        base_url = (config.base_url or "http://localhost:11434/v1").rstrip("/")
        self._client = AsyncOpenAI(
            api_key=config.api_key or "sk-no-key-required",
            base_url=base_url,
        )

    @property
    def name(self) -> str:
        return "Custom (OpenAI-compatible)"

    def is_configured(self) -> bool:
        return bool(self._config.base_url)

    async def chat(
        self,
        messages: List[ChatMessage],
    ) -> AsyncGenerator[str, None]:
        stream = await self._client.chat.completions.create(
            model=self._config.model,
            messages=[{"role": m.role, "content": m.content} for m in messages],
            max_tokens=self._config.max_tokens,
            temperature=self._config.temperature,
            stream=True,
        )

        async for chunk in stream:
            delta = chunk.choices[0].delta.content
            if delta:
                yield delta

File: chatbot/backend/src/llm/custom.ts
import { LLMProvider } from './interface'
import { LLMConfig, ChatMessage } from '../config'
import { parseOpenAIStream } from './openai'

/**
 * Custom OpenAI-compatible provider.
 *
 * Supports any API that follows the OpenAI chat completions format, including:
 * - Ollama (http://localhost:11434/v1)
 * - vLLM
 * - LocalAI
 * - Azure OpenAI
 * - Together AI
 * - Any OpenAI-compatible endpoint
 *
 * Implements LLMProvider directly and reuses the standalone parseOpenAIStream function.
 */
export class CustomProvider implements LLMProvider {
  readonly name = 'Custom (OpenAI-compatible)'

  isConfigured(config: LLMConfig): boolean {
    return !!config.baseUrl
  }

  async chat(
    messages: ChatMessage[],
    config: LLMConfig
  ): Promise<ReadableStream<string>> {
    if (!this.isConfigured(config)) {
      throw new Error(
        'Custom provider base URL is not configured. Set CUSTOM_API_BASE_URL in wrangler.toml or env vars.'
      )
    }

    const body = {
      model: config.model,
      messages: messages.map((m) => ({
        role: m.role,
        content: m.content,
      })),
      max_tokens: config.maxTokens,
      temperature: config.temperature,
      stream: true,
    }

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    // Add API key if provided (optional for local models like Ollama)
    if (config.apiKey) {
      headers['Authorization'] = `Bearer ${config.apiKey}`
    }

    const baseUrl = config.baseUrl!
    const response = await fetch(
      `${baseUrl.replace(/\/$/, '')}/chat/completions`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
      }
    )

    if (!response.ok) {
      const error = await response.text()
      throw new Error(`Custom provider API error (${response.status}): ${error}`)
    }

    if (!response.body) {
      throw new Error('Custom provider response body is null')
    }

    // Reuse the standalone OpenAI SSE parser — the format is identical
    return parseOpenAIStream(response.body as ReadableStream<Uint8Array>)
  }
}
